service/worker/archiver/archiver.go

```diff
@@ -0,0 +1,166 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package archiver
+
+import (
+	"time"
+
+	"github.com/uber-common/bark"
+	"github.com/uber/cadence/common/logging"
+	"github.com/uber/cadence/common/metrics"
+	"go.uber.org/cadence"
+	"go.uber.org/cadence/workflow"
+)
+
+type (
+	// Archiver is used to process archival requests
+	Archiver interface {
+		Start()
+		Finished() []uint64
+	}
+
+	archiver struct {
+		ctx           workflow.Context
+		logger        bark.Logger
+		metricsClient metrics.Client
+		concurrency   int
+		requestCh     workflow.Channel
+		resultCh      workflow.Channel
+	}
+)
+
+// NewArchiver returns a new Archiver
+func NewArchiver(
+	ctx workflow.Context,
+	logger bark.Logger,
+	metricsClient metrics.Client,
+	concurrency int,
+	requestCh workflow.Channel,
+) Archiver {
+	return &archiver{
+		ctx:           ctx,
+		logger:        logger,
+		metricsClient: metricsClient,
+		concurrency:   concurrency,
+		requestCh:     requestCh,
+		resultCh:      workflow.NewBufferedChannel(ctx, concurrency),
```

Comment:
Do you really need buffered channel here?

Reply:
I prefer the option of Unbuffered channel to future because there are several coroutines writing to this resultsCh. If we wanted to use futures I think we would need to use a selector and have a future added to the selector per coroutine.